test/starknet/MerkleProof.test.ts

```diff
@@ -0,0 +1,89 @@
+import { StarknetContract } from 'hardhat/types/runtime';
+import { expect } from 'chai';
+import { ethers, starknet } from 'hardhat';
+import { computeHashOnElements } from 'starknet/dist/utils/hash';
+import { MerkleTree } from '../shared/merkle';
+
+describe('Merkle:', () => {
+  let testMerkle: StarknetContract;
+
+  before(async function () {
+    this.timeout(800000);
+    const testMerkleFactory = await starknet.getContractFactory(
+      './contracts/starknet/TestContracts/Test_Merkle.cairo'
+    );
+    testMerkle = await testMerkleFactory.deploy();
+  });
+
+  it('The library should verify a merkle proof for a given root', async () => {
+    const values = [];
+    // Generating random data for the merkle tree
+    for (let i = 0; i < 100; i++) {
+      values.push([
+        ethers.Wallet.createRandom().address,
+        ethers.utils.hexlify(ethers.utils.randomBytes(1)),
+      ]);
+    }
+
+    // computing the hash of each address value pair, and sorting
+    const merkleData = values
+      .map((v, i) => [computeHashOnElements(v), v[0], v[1]])
+      .sort(function (a, b) {
+        if (a > b) return 1;
+        if (a < b) return -1;
```

Comment:
why are we sorting the values here? this will make the hashes sorted, but will destroy the ordering required to recompute the hashes, won't it?

Reply:
the hash of each leaf is just `computeHashOnElements([address, vp.low, vp.high]), so sorting doesnt affect them.

Reply:
I guess I'm missing something.
The way I view it, we have a tree, and build a list of hashes, from the bottom to the top, of the required hashes in order to recompute the root, given a leaf.
```
        [8]
       /     \
   [6]       [7]
  /     \     /    \
[2] [3] [4] [5]
```

E.g: if I want to prove [3], then the proofs will be [2, 7]. The verifier then simply needs to hash [2, 3] to get [6], and then has [6,7] to get [8].

If we sort the array `[2, 7]`, this won't work anymore. So I guess I'm missing something: or maybe I'm completely misunderstanding what we're even doing here! ^^

Reply:
sorry missed this. 

Whats being sorted ^ there is the leaf nodes, not the proof. The list of hashes does not include the ones at higher levels (6,7,8 in that example)

Reply:
Ah indeed my bad, I thought I was looking at the `getProof` implementation haha. I understand now the quiproquo.
All fine ser